src/Storages/ColumnsDescription.cpp

```diff
@@ -869,6 +870,63 @@ bool ColumnsDescription::hasCompressionCodec(const String & column_name) const
     return it != columns.get<1>().end() && it->codec != nullptr;
 }
 
+void ColumnsDescription::setDefaultCompressionCodec(const String & codec_name)
+{
+    auto columns_to_override = std::invoke([this, &codec_name](){    
+        std::vector<std::pair<String, String>> ret;
+        String codec_info_str = std::invoke([&codec_name]{
+            WriteBufferFromOwnString codec_buf;
+            writeCString("\tCODEC(", codec_buf);
+            writeCString(codec_name.c_str(), codec_buf);
+            writeCString(")\n", codec_buf);
+            return codec_buf.str();
+        });
+
+        for(const auto & column : columns)
+        {
+            if(column.codec) continue;
+
+            String new_column_desc = std::invoke([&column, &codec_info_str]{
+                WriteBufferFromOwnString buf;
+                column.writeText(buf);
+                auto first_part = buf.str();
+                // Remove the unnecessary new line
+                first_part.pop_back();
+
+                WriteBufferFromOwnString whole_part;        
+                writeCString(first_part.c_str(), whole_part);
+                writeCString(codec_info_str.c_str(), whole_part);
+                return whole_part.str();
+            });
```

Comment:
Here you assume that the codec is the last part for the column declaration, but that is [not true](https://clickhouse.com/docs/en/sql-reference/statements/create/table#with-explicit-schema). There can be a TTL expression, or even a [column level setting](https://clickhouse.com/docs/en/sql-reference/statements/create/table#with-explicit-schema) or statistics. Check the [implementation](https://github.com/ClickHouse/ClickHouse/blob/9c10d4f402b4c83c250a5bb4367aaa24142a21e6/src/Storages/ColumnsDescription.cpp#L145-L168) of `ColumnDescription::writeText`.